pkg/migration-tool/writer/shards.go

```diff
@@ -0,0 +1,178 @@
+package writer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/migration-tool/utils"
+)
+
+// shardsSet represents a set of shards. It consists of configuration related
+// to handling and management of shards.
+type shardsSet struct {
+	num         int
+	set         []*shard
+	cancelFuncs []context.CancelFunc
+
+	done    chan struct{}
+	errChan chan error
+}
+
+// newShardsSet creates a shards-set and initializes it. It creates independent clients for each shard,
+// contexts and starts the shards. The shards listens to the writer routine, which is responsible for
+// feeding the shards with data blocks for faster (due to sharded data) flushing.
+func newShardsSet(writerCtx context.Context, writeURL string, numShards int) (*shardsSet, error) {
+	var (
+		set         = make([]*shard, numShards)
+		cancelFuncs = make([]context.CancelFunc, numShards)
+	)
+	ss := &shardsSet{
+		num:     numShards,
+		done:    make(chan struct{}, numShards),
+		errChan: make(chan error, numShards),
+	}
+	for i := 0; i < numShards; i++ {
+		client, err := utils.NewClient(fmt.Sprintf("writer-shard-%d", i), writeURL, utils.Write, model.Duration(defaultWriteTimeout))
+		if err != nil {
+			return nil, fmt.Errorf("creating write-shard-client-%d: %w", i, err)
+		}
+		ctx, cancelFunc := context.WithCancel(writerCtx)
+		shard := &shard{
+			ctx:          ctx,
+			client:       client,
+			queue:        make(chan *[]prompb.TimeSeries),
+			copyShardSet: ss,
+		}
+		cancelFuncs[i] = cancelFunc
+		set[i] = shard
+	}
+	ss.set = set
+	ss.cancelFuncs = cancelFuncs
+	// Run the shards.
+	for i := 0; i < ss.num; i++ {
+		go ss.set[i].run(i)
+	}
+	return ss, nil
+}
+
+// scheduleTS batches the time-series based on the available shards. These batches are then
+// fed to the shards in a single go which then consume and push to the respective clients.
+func (s *shardsSet) scheduleTS(ts []prompb.TimeSeries) {
+	var (
+		tsIndex     = 0
+		tsEachShard = len(ts) / s.num
+	)
+	for shardIndex := tsIndex; shardIndex < s.num; shardIndex++ {
+		var batch []prompb.TimeSeries
+		if shardIndex == s.num-1 {
+			batch = ts[tsIndex:]
+		} else {
+			batch = ts[tsIndex : tsIndex+tsEachShard]
```

Comment:
Prometheus shards items using a consistent hash on the series. This seems to round-robin. I think it's better to use consistent hash like Prometheus since long term stores may optimize for that.

Reply:
Won't that just add to the cost, as I am unable to understand the importance of it than just dividing among shards for which here we use indexes.